File: algorithms/medium/minimum_time_to_make_rope_colorful.ex
defmodule Solution do
  @spec min_cost(colors :: String.t(), needed_time :: [integer]) :: integer
  def min_cost(colors, needed_time) do
    colors
    |> String.graphemes()
    |> Enum.zip(needed_time)
    |> Enum.reduce({"", 0, 0, 0}, fn {c, n}, {prev, ans, sum, ma} ->
      if c != prev do
        {c, ans + sum - ma, n, n}
      else
        {c, ans, sum + n, max(ma, n)}
      end
    end)
    |> then(fn {_, ans, sum, ma} ->
      ans + sum - ma
    end)
  end
end

File: algorithms/medium/minimum_time_to_make_rope_colorful.rs
impl Solution {
    pub fn min_cost(colors: String, needed_time: Vec<i32>) -> i32 {
        let colors = colors.as_bytes();
        let mut ans = 0;
        let mut sum = needed_time[0];
        let mut max = needed_time[0];
        for i in 1..colors.len() {
            if colors[i] != colors[i - 1] {
                ans += (sum - max);
                sum = 0;
                max = 0;
            }
            sum += needed_time[i];
            max = max.max(needed_time[i]);
        }
        ans + sum - max
    }
}
